frontends/p4/alias.h

```diff
@@ -0,0 +1,255 @@
+        return result;
+    }
+
+    /// Append suffix to each location in the set
+    const SetOfLocations* append(cstring suffix) const {
+        auto result = new SetOfLocations();
+        for (auto p : paths) {
+            auto append = p->append(suffix);
+            result->add(append);
+        }
+        return result;
+    }
+
+    void dbprint(std::ostream& out) const override {
+        for (auto p : paths)
+            out << p << std::endl;
+    }
+};
+
+/// Computes the SetOfLocations read and written by an expression.
+class ReadsWrites : public Inspector {
+    const ReferenceMap* refMap;
+    std::map<const IR::Expression*, const SetOfLocations*> rw;
+    bool noMethodCalls;  /// If this flag is true we do not expect to see
+    /// method calls in the expressions - except calls to isValid().
+
+ public:
+    ReadsWrites(const ReferenceMap* refMap, bool noMethodCalls) :
+            refMap(refMap), noMethodCalls(noMethodCalls)
+    { setName("ReadsWrites"); }
+
+    void postorder(const IR::Operation_Binary* expression) override {
+        auto left = ::get(rw, expression->left);
+        auto right = ::get(rw, expression->right);
+        rw.emplace(expression, left->join(right));
+    }
+
+    void postorder(const IR::PathExpression* expression) override {
+        auto decl = refMap->getDeclaration(expression->path);
+        auto path = new LocationPath(decl);
+        auto locs = new SetOfLocations(path);
+        rw.emplace(expression, locs);
+    }
+
+    void postorder(const IR::Operation_Unary* expression) override {
+        auto e = ::get(rw, expression->expr);
+        rw.emplace(expression, e);
+    }
+
+    void postorder(const IR::Member* expression) override {
+        auto e = ::get(rw, expression->expr);
+        auto result = e->append(expression->member);
+        rw.emplace(expression, result);
+    }
+
+    void postorder(const IR::ArrayIndex* expression) override {
+        auto e = ::get(rw, expression->left);
+        const SetOfLocations* result;
+        if (expression->right->is<IR::Constant>()) {
+            int index = expression->right->to<IR::Constant>()->asInt();
+            result = e->append(Util::toString(index));
+        } else {
+            result = e->append("*");
+        }
+        rw.emplace(expression, result);
+    }
+
+    void postorder(const IR::Literal* expression) override {
+        rw.emplace(expression, new SetOfLocations());
+    }
+
+    void postorder(const IR::TypeNameExpression* expression) override {
+        rw.emplace(expression, new SetOfLocations());
+    }
+
+    void postorder(const IR::Operation_Ternary* expression) override {
+        auto e0 = ::get(rw, expression->e0);
+        auto e1 = ::get(rw, expression->e1);
+        auto e2 = ::get(rw, expression->e2);
+        rw.emplace(expression, e0->join(e1)->join(e2));
+    }
+
+    void postorder(const IR::Slice* expression) override {
+        auto e = ::get(rw, expression->expr);
```

Comment:
I'm confused how this even compiles?  IR::Slice has no field called `expr`...

Reply:
well, it didn't...
sorry about that.